pkg/tun/tun_darwin.go

```diff
@@ -0,0 +1,246 @@
+package tun
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"os"
+	"runtime"
+	"unsafe"
+
+	"golang.org/x/net/ipv4"
+	"golang.org/x/net/ipv6"
+	"golang.org/x/sys/unix"
+
+	"github.com/telepresenceio/telepresence/v2/pkg/tun/buffer"
+)
+
+const sysProtoControl = 2
+const uTunOptIfName = 2
+const uTunControlName = "com.apple.net.utun_control"
+
+type Device struct {
+	*os.File
+	name string
+}
+
+// OpenTun creates a new TUN device and ensures that it is up and running.
+func OpenTun() (*Device, error) {
+	fd, err := unix.Socket(unix.AF_SYSTEM, unix.SOCK_DGRAM, sysProtoControl)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open DGRAM socket: %v", err)
+	}
+	defer func() {
+		if err != nil {
+			_ = unix.Close(fd)
+		}
+	}()
```

Comment:
1. Probably need to close it even if `err != nil`?
2. Does the daemon shell out for anything?  If it might fork, then when dealing with raw file descriptors we need to set up close-on-exec and also mess with `syscall.ForkClock`

```go
	syscall.ForkLock.RLock()
	fd, err := unix.Socket(unix.AF_SYSTEM, unix.SOCK_DGRAM, sysProtoControl)
	if err == nil {
		syscall.CloseOnExec(fd)
	}
	syscall.ForkLock.RUnlock()
```

Or, on platforms with `unix.SOCK_CLOEXEC` (IDK if macOS does; GOOS=linux has it, but depends on Linux 2.6.27 (2008-10-09), which is slightly newer than Go itself depends on, 2.6.23 (2007-10-09), so maybe need to check for EINVAL and try again without it?  Or just declare that we need 2.6.27?).

```go
	fd, err := unix.Socket(unix.AF_SYSTEM, unix.SOCK_DGRAM|unix.SOCK_CLOEXEC, 0)
```

See also:
- https://git.lukeshu.com/go/libsystemd/tree/sd_daemon/notify_linux.go#n93
- https://git.lukeshu.com/go/libsystemd/tree/sd_daemon/notify_nonlinux.go#n93
- https://github.com/golang/go/blob/go1.15.12/src/net/sock_cloexec.go#L18-L50

Reply:
The close must only happen if `err != nil` because `fd` must remain open on success. I use this construct instead of repeatedly do a `unix.Close(fd)` on all error returns that follow.

Reply:
Gotcha, I was thinking that `os.NewFile` `dup(2)`'d the fd, but I was mixing it up with `net.FileConn`.